File: solutions/JavaScript/permutations_used_array.js
/**
 * Generate all permutations using backtracking with used array.
 * Time: O(n! * n), Space: O(n!) for result
 * @param {number[]} nums
 * @return {number[][]}
 */
function permutationsUsedArray(nums) {
    const result = [];
    const used = new Array(nums.length).fill(false);
    const current = [];

    function backtrack() {
        // Base case: we've used all numbers
        if (current.length === nums.length) {
            result.push([...current]);
            return;
        }

        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue;

            // Choose
            current.push(nums[i]);
            used[i] = true;
            // Explore
            backtrack();
            // Unchoose
            current.pop();
            used[i] = false;
        }
    }

    backtrack();
    return result;
}

console.log(permutationsUsedArray([1, 2, 3]));
// Output: [[1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]]

File: solutions/Rust/permutations_used_array.rs
pub fn permutations_used_array(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut result = vec![];
    let mut used = vec![false; nums.len()];
    let mut current = vec![];
    backtrack(&nums, &mut used, &mut current, &mut result);
    result
}

fn backtrack(nums: &Vec<i32>, used: &mut Vec<bool>, current: &mut Vec<i32>, result: &mut Vec<Vec<i32>>) {
    // Base case: we've used all numbers
    if current.len() == nums.len() {
        result.push(current.clone());
        return;
    }

    for i in 0..nums.len() {
        if used[i] {
            continue;
        }

        // Choose
        current.push(nums[i]);
        used[i] = true;
        // Explore
        backtrack(nums, used, current, result);
        // Unchoose
        current.pop();
        used[i] = false;
    }
}

fn main() {
    let result = permutations_used_array(vec![1, 2, 3]);
    println!("{:?}", result);
    // Output: [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
}
